File: src/main/java/mnm/mods/tabbychat/ChatManager.java
package mnm.mods.tabbychat;

import com.google.common.collect.ImmutableSet;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.google.gson.reflect.TypeToken;
import com.mojang.authlib.GameProfile;
import mnm.mods.tabbychat.api.Channel;
import mnm.mods.tabbychat.api.Chat;
import mnm.mods.tabbychat.api.UserChannel;
import mnm.mods.tabbychat.api.events.MessageAddedToChannelEvent;
import mnm.mods.tabbychat.settings.ServerSettings;
import mnm.mods.tabbychat.settings.TabbySettings;
import mnm.mods.tabbychat.util.ChannelTypeAdapter;
import mnm.mods.tabbychat.util.ChatTextUtils;
import mnm.mods.tabbychat.util.DateTimeTypeAdapter;
import mnm.mods.util.config.ValueMap;
import mnm.mods.util.text.TextBuilder;
import net.minecraft.client.Minecraft;
import net.minecraft.util.EnumTypeAdapterFactory;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.Style;
import net.minecraft.util.text.TextFormatting;
import net.minecraftforge.common.MinecraftForge;
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.reflect.TypeUtils;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;

public class ChatManager implements Chat {

    public static final int MAX_CHAT_LENGTH = 256;

    private static ChatManager instance;

    public static final AbstractChannel DEFAULT_CHANNEL = new AbstractChannel("*") {
        @Override
        public String getName() {
            return getAlias();
        }

        @Override
        public String toString() {
            return "#" + getAlias();
        }

        // Don't mess with this channel
        @Override
        public void setAlias(String alias) {
        }

        @Override
        public void setPrefix(String prefix) {
        }

        @Override
        public void setPrefixHidden(boolean hidden) {
        }

        @Override
        public void setCommand(String command) {
        }
    };

    private Gson gson = new GsonBuilder()
            .excludeFieldsWithoutExposeAnnotation()
            .registerTypeHierarchyAdapter(ITextComponent.class, new ITextComponent.Serializer())
            .registerTypeAdapter(Style.class, new Style.Serializer())
            .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
            .registerTypeAdapter(Channel.class, new ChannelTypeAdapter(this))
            .registerTypeAdapter(LocalDateTime.class, new DateTimeTypeAdapter())
            .create();

    private Map<String, Channel> allChannels = new HashMap<>();
    private Map<GameProfile, UserChannel> allPms = new HashMap<>();

    private Set<Channel> channels = new HashSet<>();
    private Map<Channel, List<ChatMessage>> messages = new HashMap<>();
    private Map<Channel, List<ChatMessage>> msgsplit = new HashMap<>();

    private ServerSettings server() {
        return TabbyChatClient.getInstance().getServerSettings();
    }

    private TabbySettings settings() {
        return TabbyChatClient.getInstance().getSettings();
    }

    private ChatManager() {
    }

    public static ChatManager instance() {
        if (instance == null) {
            instance = new ChatManager();
        }
        return instance;
    }

    @Override
    public Channel getChannel(String name) {
        return allChannels.computeIfAbsent(name, this.getChannel(name, server().channels, ChatChannel::new));
    }

    @Override
    public UserChannel getUserChannel(GameProfile gameProfile) {
        return allPms.computeIfAbsent(gameProfile, this.getChannel(gameProfile.getName(), server().pms, DirectChannel::new));
    }

    private <T, U> Function<U, T> getChannel(String name, ValueMap<T> config, Function<U, T> absent) {
        return profile -> config.get().computeIfAbsent(name, k -> absent.apply(profile));
    }

    /**
     * Parses a channel name and returns the channel. format must start with either # or @
     *
     * @param format The representation of the channel as a string
     * @return The channel
     */
    public Optional<Channel> parseChannel(String format) {
        if (format.length() <= 1) {
            return Optional.empty();
        }
        String name = format.substring(1);
        switch (format.charAt(0)) {
            case '@':
                return Optional.of(getUserChannel(new GameProfile(null, name)));
            case '#':
                return Optional.of(getChannel(name));
            default:
                throw new IllegalArgumentException();
        }

    }

    @Deprecated
    public void addChannel(AbstractChannel channel) {
        channels.add(channel);
    }

    @Deprecated
    public void removeChannel(AbstractChannel channel) {
        channels.remove(channel);
    }

    @Override
    public Set<Channel> getChannels() {
        // TODO only show visible channels.
        return ImmutableSet.copyOf(channels);
    }

    @Override
    public List<ChatMessage> getMessages(Channel channel) {
        return Collections.unmodifiableList(getChannelMessages(channel));
    }

    private List<ChatMessage> getChannelMessages(Channel channel) {
        return this.messages.computeIfAbsent(channel, k -> new ArrayList<>());
    }

    public List<ChatMessage> getVisible(Channel channel, int width) {
        return this.msgsplit.computeIfAbsent(channel, a -> ChatTextUtils.split(getMessages(a), width));
    }

    public void markDirty(Channel channel) {
        if (channel == null) {
            this.msgsplit.clear();
        } else {
            this.msgsplit.remove(channel);
        }
    }

    @Override
    public void addMessage(Channel channel, ITextComponent message) {
        addMessage(channel, message, 0);
    }

    public void addMessage(Channel channel, ITextComponent text, int id) {
        MessageAddedToChannelEvent event = new MessageAddedToChannelEvent.Pre(text.deepCopy(), id, channel);
        if (MinecraftForge.EVENT_BUS.post(event) || event.getText() == null) {
            return;
        }
        text = event.getText();
        id = event.getId();

        if (id != 0) {
            removeMessages(id);
        }

        int uc = Minecraft.getInstance().ingameGUI.getTicks();
        ChatMessage msg = new ChatMessage(uc, text, id, true);
        List<ChatMessage> messages = getChannelMessages(channel);
        messages.add(0, msg);

        trimMessages(messages, settings().advanced.historyLen.get());

        save();

        MinecraftForge.EVENT_BUS.post(new MessageAddedToChannelEvent.Post(text, id, channel));

        markDirty(channel);
    }

    private void trimMessages(List<?> list, int size) {
        Iterator<?> iter = list.iterator();

        for (int i = 0; iter.hasNext(); i++) {
            iter.next();
            if (i > size) {
                iter.remove();
            }
        }
    }

    public void removeMessages(int id) {
        for (List<ChatMessage> messages : this.messages.values()) {
            messages.removeIf(m -> m.getID() == id);
        }
        save();
    }

    public void removeMessageAt(AbstractChannel channel, int index) {
        getChannelMessages(channel).remove(index);
        save();
    }

    public void clearMessages() {
        this.messages.clear();
    }

    private static final Object lock = new Object();

    public void loadFrom(Path dir) throws IOException {
        synchronized (lock) {
            loadFrom_(dir);
        }
    }

    private synchronized void loadFrom_(Path dir) throws IOException {
        Path file = dir.resolve("data.gz");
        if (Files.notExists(file)) {
            return;
        }
        try (Reader gzin = new InputStreamReader(new GzipCompressorInputStream(Files.newInputStream(file)), StandardCharsets.UTF_8)) {
            JsonObject root = gson.fromJson(gzin, JsonObject.class);

            clearMessages();
            allChannels.clear();
            allPms.clear();

            Type type = new TypeToken<List<ChatMessage>>() {
            }.getType();

            List<ChatMessage> def = gson.fromJson(root.get("default"), type);
            getChannelMessages(DEFAULT_CHANNEL).addAll(def);

            readChannels(root.get("chans").getAsJsonObject());

            // active channels
            JsonObject active = root.get("active").getAsJsonObject();
            JsonArray achans = active.get("chans").getAsJsonArray();
            for (JsonElement e : achans) {
                addChannel(gson.fromJson(e, AbstractChannel.class));
            }

            String time;
            if (root.has("datetime")) {
                Instant datetime = Instant.ofEpochSecond(root.get("datetime").getAsLong());
                time = datetime.toString();
            } else {
                time = "UNKNOWN";
            }
            ITextComponent chat = new TextBuilder()
                    .text("Chat log from " + time)
                    .format(TextFormatting.GRAY)
                    .build();

            for (Channel c : getChannels()) {
                if (!getMessages(c).isEmpty()) {
                    addMessage(c, chat, -1);
                }
            }
        }
    }

    private void readChannels(JsonObject obj) {
        for (Entry<String, JsonElement> entry : obj.entrySet()) {
            Channel chan = getChannel(entry.getKey());
            List<ChatMessage> list = gson.fromJson(entry.getValue(), new TypeToken<List<ChatMessage>>(){}.getType());
            getChannelMessages(chan).addAll(list);
        }
    }

    private void readUserChannels(JsonObject obj) {
        for (Entry<String, JsonElement> entry : obj.entrySet()) {
            // profile is unsafe, but I'm not sent the uuid. Nothing I can do
            GameProfile profile = new GameProfile(null, entry.getKey());
            Channel chan = getUserChannel(profile);
            Type type = TypeUtils.parameterize(List.class, ChatMessage.class);
            List<ChatMessage> list = gson.fromJson(entry.getValue(), type);
            getChannelMessages(chan).addAll(list);
        }
    }

    public void save() {
        synchronized (lock) {
            try {
                saveTo(server().getPath().getParent());
            } catch (IOException e) {
                TabbyChat.logger.warn("Error while saving chat data", e);
            }
        }
    }

    private synchronized void saveTo(Path dir) throws IOException {
        JsonObject root = new JsonObject();
        root.addProperty("datetime", Instant.now().getEpochSecond());
        root.add("default", gson.toJsonTree(getMessages(DEFAULT_CHANNEL)));

        JsonObject chans = new JsonObject();
        root.add("chans", chans);

        for (Channel c : messages.keySet()) {
            chans.add(c.toString(), gson.toJsonTree(getMessages(c)));
        }

        // active channels
        JsonObject active = new JsonObject();
        root.add("active", active);

        JsonArray achans = new JsonArray();
        active.add("chans", achans);

        for (Channel c : channels) {
            if (c == DEFAULT_CHANNEL) {
                continue;
            }
            achans.add(gson.toJson(c));
        }

        Path file = dir.resolve("data.gz");
        Files.createDirectories(dir);
        try (Writer gzout = new OutputStreamWriter(new GzipCompressorOutputStream(Files.newOutputStream(file)), StandardCharsets.UTF_8)) {
            gson.toJson(root, gzout);
        }
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/ChatManager.kt
package mnm.mods.tabbychat.client

import com.google.common.collect.ImmutableSet
import com.google.gson.GsonBuilder
import mnm.mods.tabbychat.CHATBOX
import mnm.mods.tabbychat.TabbyChat
import mnm.mods.tabbychat.api.Channel
import mnm.mods.tabbychat.api.ChannelType
import mnm.mods.tabbychat.api.Chat
import mnm.mods.tabbychat.api.events.MessageAddedToChannelEvent
import mnm.mods.tabbychat.client.gui.ChatBox
import mnm.mods.tabbychat.client.settings.ServerSettings
import mnm.mods.tabbychat.client.settings.TabbySettings
import mnm.mods.tabbychat.util.*
import mnm.mods.tabbychat.util.config.ConfigEvent
import net.minecraft.util.EnumTypeAdapterFactory
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.StringTextComponent
import net.minecraft.util.text.Style
import net.minecraft.util.text.TextFormatting
import net.minecraftforge.common.MinecraftForge
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream
import java.io.*
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.time.LocalDateTime
import java.util.*

object ChatManager : Chat {

    private val gson = GsonBuilder()
            .registerTypeHierarchyAdapter(ITextComponent::class.java, ITextComponent.Serializer())
            .registerTypeAdapter(Style::class.java, Style.Serializer())
            .registerTypeAdapterFactory(EnumTypeAdapterFactory())
            .registerTypeHierarchyAdapter(Channel::class.java, ToStringAdapter {
                this.parseChannel(it) ?: throw IOException("Serialized channels must start with @ or #")
            })
            .registerTypeAdapter(LocalDateTime::class.java, ToStringAdapter(LocalDateTime::parse))
            .create()

    const val MAX_CHAT_LENGTH = 256

    private val lock = Any()


    private val allChannels: MutableMap<String, Channel> = hashMapOf()
    private val allPms: MutableMap<String, Channel> = hashMapOf()

    /**
     * A map of all the messages per channel
     */
    private val messages = HashMap<Channel, MutableList<ChatMessage>>()

    /**
     * The split messages used for rendering the chatbox.
     */
    private val msgsplit = HashMap<Channel, List<ChatMessage>>()

    override val channels: Set<Channel>
        get() = ImmutableSet.copyOf<Channel>(ChatBox.getChannels())

    private val server: ServerSettings
        get() = TabbyChatClient.serverSettings

    private val settings: TabbySettings
        get() = TabbyChatClient.settings

    init {
        allChannels["*"] = DefaultChannel
        val dirtys = listOf(
                settings.general::timestampChat,
                settings.general::timestampColor,
                settings.general::timestampStyle
        )
        MinecraftForge.EVENT_BUS.listen<ConfigEvent> {
            if (it.prop in dirtys) {
                markDirty()
            }
        }
    }

    override fun getChannel(type: ChannelType, name: String): Channel {
        return allChannels.getOrPut(name) {
            server.channels.associateBy { it.name }.getOrElse(name) {
                val channels = server.channels.toMutableList()
                val channel = ChannelImpl(type, name)
                if (type === ChannelType.USER) {
                    channel.prefix = "/msg $name"
                }
                channels.add(channel)
                server.save()
                channel
            }
        }
    }

    /**
     * Parses a channel name and returns the channel. format must start with either # or @
     *
     * @param format The representation of the channel as a string
     * @return The channel
     */
    fun parseChannel(format: String): Channel? {
        if (format.isEmpty()) {
            return null
        }
        val name = format.substring(1)
        return when (format[0]) {
            '@' -> getChannel(ChannelType.USER, name)
            '#' -> getChannel(ChannelType.CHAT, name)
            '*' -> DefaultChannel
            else -> null
        }

    }

    fun coerceChannelName(name: String): String {
        if (name.isBlank()) {
            return "#"
        }
        val firstChar = name.first()
        if (firstChar != '#' && firstChar != '@') {
            return "#$name"
        }
        return name
    }

    override fun getMessages(channel: Channel): List<ChatMessage> {
        return Collections.unmodifiableList(getChannelMessages(channel))
    }

    private fun getChannelMessages(channel: Channel): MutableList<ChatMessage> {
        return messages.getOrPut(channel) { ArrayList() }
    }

    fun getVisible(channel: Channel, width: Int): List<ChatMessage> {
        return msgsplit.getOrPut(channel) { ChatTextUtils.split(getMessages(channel), width) }
    }

    fun markDirty(channel: Channel? = null) {
        if (channel == null) {
            this.msgsplit.clear()
        } else {
            this.msgsplit.remove(channel)
        }
    }

    override fun addMessage(channel: Channel, message: ITextComponent) {
        addMessage(channel, message, 0)
    }

    fun addMessage(channel: Channel, text: ITextComponent, id: Int) {
        val event = MessageAddedToChannelEvent.Pre(text.deepCopy(), id, channel)
        if (MinecraftForge.EVENT_BUS.post(event) || event.text == null) {
            return
        }
        val text = event.text!!
        val id = event.id

        if (id != 0) {
            removeMessages(channel, id)
        }

        val uc = mc.ingameGUI.ticks
        val msg = ChatMessage(uc, text, id)
        val messages = getChannelMessages(channel)
        messages.add(0, msg)

        trimMessages(messages, settings.advanced.historyLen)

        MinecraftForge.EVENT_BUS.post(MessageAddedToChannelEvent.Post(text, id, channel))

        save()
        markDirty(channel)
    }

    private fun trimMessages(list: MutableList<*>, size: Int) {
        val iter = list.iterator()

        var i = 0
        while (iter.hasNext()) {
            iter.next()
            if (i > size) {
                iter.remove()
            }
            i++
        }
    }

    fun removeMessages(chan: Channel, id: Int) {
        getChannelMessages(chan).removeIf { m -> m.id == id }
    }

    fun removeMessageAt(channel: Channel, index: Int) {
        getChannelMessages(channel).removeAt(index)
        save()
    }

    fun clearMessages() {
        this.messages.clear()
        this.msgsplit.clear()

        this.allChannels.clear()
        this.allChannels["*"] = DefaultChannel
        this.allPms.clear()

        ChatBox.clearMessages()
    }

    @Throws(IOException::class)
    fun loadFrom(dir: Path) {
        synchronized(lock) {
            loadFromPrivate(dir)
        }
    }

    @Synchronized
    @Throws(IOException::class)
    private fun loadFromPrivate(dir: Path) {
        val file = dir.resolve("data.gz")
        if (Files.notExists(file)) {
            return
        }

        clearMessages()

        newCompressedReader(file).use { gzin ->
            val root = gson.fromJson(gzin, PersistantChat::class.java)

            for ((k, v) in root.messages) {
                this.messages[k] = v.toMutableList()
            }
            ChatBox.addChannels(root.active)

            val chat: ITextComponent = StringTextComponent("Chat log from ${root.time}").style {
                color = TextFormatting.GRAY
            }

            for (c in channels) {
                if (getMessages(c).isNotEmpty()) {
                    addMessage(c, chat, -1)
                }
            }
        }
    }

    fun save() {
        synchronized(lock) {
            try {
                saveTo(server.config.nioPath.parent)
            } catch (e: IOException) {
                TabbyChat.logger.warn(CHATBOX, "Error while saving chat data", e)
            }

        }
    }

    @Synchronized
    @Throws(IOException::class)
    private fun saveTo(dir: Path) {
        val file = dir.resolve("data.gz")
        Files.createDirectories(dir)

        newCompressedWriter(file).use {
            gson.toJson(PersistantChat(messages, ChatBox.getChannels()), PersistantChat::class.java, it)
        }
    }

    private class PersistantChat(
            messages: Map<Channel, List<ChatMessage>>?,
            active: Collection<Channel>?,
            datetime: LocalDateTime? = LocalDateTime.now()
    ) {
        val messages = messages ?: emptyMap()
        val active = active ?: emptyList()
        val time = datetime?.toString() ?: "UNKNOWN"
    }

    @Throws(IOException::class)
    private fun newCompressedReader(path: Path): Reader {
        return InputStreamReader(GzipCompressorInputStream(Files.newInputStream(path)), StandardCharsets.UTF_8)
    }

    @Throws(IOException::class)
    private fun newCompressedWriter(path: Path): Writer {
        return OutputStreamWriter(GzipCompressorOutputStream(Files.newOutputStream(path)), StandardCharsets.UTF_8)
    }

}
